docs: clarify different types of API access (#535)
canonical/pebble: docs/explanation/api-and-clients.md

The Pebble daemon exposes an API (HTTP over a unix socket) to allow remote clients to interact with the daemon. It can start and stop services, add configuration layers to the plan, and so on.

> If `pebble run` is started with the `--http <address>` option, Pebble also exposes open-access HTTP endpoints using the given TCP address (see {ref}`api-access-levels` below).

Reviewer: Can we also change the help text from `pebble run --help` from [this](https://github.com/canonical/pebble/blob/3c8d2b6affe94b5428e6e04fcd21092ed50443d7/internals/cli/cmd_run.go#L60):

```
	"--http":        `Start HTTP API listening on this address (e.g., ":4000")`,
```

to something like:

```
	"--http":        `Start HTTP API listening on this address (e.g., ":4000") and expose open-access endpoints`,
```

Author: Good suggestion. I've added another commit with this change

API endpoints fall into one of three access levels, from least restricted to most restricted:

* **Open-access** - Allowed from any user, even unauthenticated users using the HTTP-over-TCP listener.

>     * `GET /v1/system-info`, which returns the Pebble version and an identifier for the daemon

Reviewer: It's actually an identifier for this specific boot. Could just say "Pebble version and other information", and leave details to Tiexin's API page.

Author: Changed - thanks!

>     * `GET /v1/health`, which returns a boolean to indicate the state of Pebble's health checks

Reviewer: To make the meaning of true clearer:

```suggestion
    * `GET /v1/health`, which returns a boolean to indicate whether Pebble's health checks are all healthy
```

Author: Changed - thanks!

* **Read-access** - Allowed from any authenticated user. For example, listing services or viewing notices.
    * All `GET` endpoints except the admin-access `GET` endpoints
    * `POST /v1/notices`, which records a custom notice

* **Admin-access** - Only allowed from admin users. For example, adding a layer or starting a service.
    * `GET /v1/files`, which pulls a file from a remote system
    * `GET /v1/tasks/{task-id}/websocket/{websocket-id}`
    * All `POST` endpoints except `POST /v1/notices` (which is read-access)

Pebble authenticates clients that connect to the socket API using peer credentials ([`SO_PEERCRED`](https://man7.org/linux/man-pages/man7/socket.7.html)) to determine the user ID (UID) of the connecting process. If this UID is 0 (root) or the UID of the Pebble daemon, the user's access level is `admin`, otherwise the access level is `read`.

If Pebble can't authenticate the user, the user's access level is `untrusted`. Unauthenticated users can only use open-access endpoints.

> Pebble doesn't try to authenticate users that connect over TCP. So any user that connects over TCP is unauthenticated (access level `untrusted`) and can only use open-access endpoints.

Reviewer: Just nothing this will change with Tiexin's upcoming "basic auth" work for the metrics endpoint, but that's a while away -- so this is correct for now.
